File: src/data/types/purchaseShipResult.ts
import type { Ship } from "./ship.ts";
import type { ShipyardPool } from "./shipyardPool.ts";
import type { Wallet } from "./wallet.ts";

// Necessary completion: Agent 20's contract names `purchaseShip(...): Ship
// | PurchaseError` but never defines PurchaseError, and a bare Ship
// return on success can't also carry the updated pool/wallet a pure
// function must return alongside it. Mirrors the existing HireResult
// pattern exactly (Phase 4's equivalent for NPC crew acquisition).
export interface PurchaseShipSucceeded {
  purchased: true;
  ship: Ship;
  updatedPool: ShipyardPool;
  updatedWallet: Wallet;
}

export interface PurchaseShipRejected {
  purchased: false;
  reason: string;
}

export type PurchaseShipResult = PurchaseShipSucceeded | PurchaseShipRejected;

File: unity/ProfitableCore/Schema/PurchaseShipResult.cs
namespace Profitable.Core.Schema;

// Ports src/data/types/purchaseShipResult.ts's PurchaseShipResult =
// PurchaseShipSucceeded | PurchaseShipRejected discriminated union,
// mirroring HireResult exactly.
public abstract class PurchaseShipResult
{
    public abstract bool Purchased { get; }
}

public sealed class PurchaseShipSucceeded : PurchaseShipResult
{
    public override bool Purchased => true;
    public Ship Ship { get; init; } = new();
    public ShipyardPool UpdatedPool { get; init; } = new();
    public Wallet UpdatedWallet { get; init; } = new();
}

public sealed class PurchaseShipRejected : PurchaseShipResult
{
    public override bool Purchased => false;
    public string Reason { get; init; } = string.Empty;
}
